libpromises/processes_select_proc.c

```diff
@@ -0,0 +1,809 @@
+
+    // a "shouldn't happen" error
+    if (!found)
+    {
+        Log(LOG_LEVEL_ERR, "error searching for uid %d in %s",
+          uid, statusfile);
+
+        return false;
+    }
+
+    JsonObjectAppendInteger(pdata, JPROC_KEY_UID, uid);
+
+    // "probably shouldn't happen"... uid of process is untranslatable into name
+    struct passwd *pwd = getpwuid(uid);
+    if (!pwd)
+    {
+        Log(LOG_LEVEL_WARNING, "could not translate uid %d into a username", uid);
+        return false;
+    }
+
+    JsonObjectAppendString(pdata, JPROC_KEY_UNAME, pwd->pw_name);
+
+    return found;
+}
+
+/*
+ * The caller is looking for a pattern such as "pts/[0-9]+".
+ * Here we examine likely places under "/dev".
+ * In this initial implementation we only look for "/dev/pts/nnn"/.
+ *
+ * The "/proc" data has given us the major/minor device "ttyn".
+ * We look through our possible candidate list in "/dev" for a match.
+ */
+static bool LoadProcTTY(JsonElement *pdata, int ttyn)
+{
+    char ttyname[CF_MAXVARSIZE];
+    struct stat statbuf;
+    int ttymajor = major(ttyn);
+    int ttyminor = minor(ttyn);
+
+    JsonObjectAppendInteger(pdata, JPROC_KEY_TTY, ttyn);
+
+    /*
+     * From kernel "devices.txt", character-major 136-143 inclusive are "pty".
+     * We do a quick sanity-check on "/dev/pts/nnnn".
+     * If it looks OK, we set the name "pts/nnn".
+     */
+    if (ttymajor >= 136 && ttymajor <= 143)
+    {
+        snprintf(ttyname, CF_MAXVARSIZE, "/dev/pts/%d", ttyminor);
+        if (stat(ttyname, &statbuf) >= 0)
+        {
+            JsonObjectAppendString(pdata, JPROC_KEY_TTYNAME, &ttyname[5]);
+            return true;
+        }
+    }
+
+    /*
+     * Could probably check for other things; "/dev/tty<n>" springs to mind.
+     * But for now, fail safely.
+     */
+
+    return false;
+}
+
+static JsonElement *LoadProcStat(pid_t pid)
+{
+    char statfile[CF_MAXVARSIZE];
+    int fd, len;
+    JsonElement *pdata;
+
+    pdata = JsonObjectCreate(12);
+
+    /* get uid from file 'status' */
+    if (! LoadProcUid(pdata, pid))
+    {
+        return pdata;
+    }
+
+    /* extract command line info and store in 'pdata' */
+    LoadProcCmd(pid, pdata);
+
+    /* open the 'stat' file */
+    snprintf(statfile, CF_MAXVARSIZE, "/%s/%d/stat", PROCDIR, pid);
+    fd = open(statfile, O_RDONLY, 0);
+    if (fd < 0) {
+        return pdata;
+    }
+
+    /* read the 'stat' file into a buffer */
+    char statbuf[CF_MAXVARSIZE];
+    len = read(fd, statbuf, CF_MAXVARSIZE -1);
+    close(fd);
+    if (len <=0) {
+        return pdata;
+    }
+
+    /*
+     * Big picture: "stat" is:
+     *    pid (some text here) S n1 n2 n3 ...
+     * We already know the pid. And the text is a variant of command (known).
+     * We are interested in the state char 'S' and numbers n1, n2, n3, ...
+     */
+
+    /* Find closing ')' and place pointer on the char 'S'. */
+    char *pt;
+    pt = strrchr(statbuf, ')');
+    pt++;
+    pt++;
```

Comment:
Here, you need to consider the worst case scenario(s):

- What if `strrchr()` didn't find a `)`? It will return `NULL`, you increment it twice, you now have a pointer to `0x000002`, which is (most likely) in kernel memory and will cause a segmentation fault (crash) when you try to use it.
- What if `)` is the last character `read()` from the file? You increment twice and `pt` is now pointing to the uninitialized part of `statbuf`.
- What if you `read()` exactly `CF_MAXVARSIZE - 1` characters from file, and the last one was a `)`? `pt` points to the last byte in your buffer, you increment twice, and it now points to something else on the stack.

You need to check that `strrchr` doesn't return NULL, and you need to make sure that your increments don't exceed `statbuf + len`, something like:

```
char *pt = strrchr(statbuf, ')');
if (pt == NULL)
{
  // Could not find closing parenthesis
  // Log some error?
  // return NULL; ?
}
size_t parenthesis_offset = pt - statbuf; // Length from statbuf start to closing parenthesis
if ((parenthesis_offset + 2) >= len)
{
  // After adding 2 to pt, it will point to uninitialized memory, 
  // either a part of statbuf that wasn't read from file or outside of statbuf
  // log error?
  // return NULL?
}
```

Reply:
Thanks!  This caused me to look beyond the se details to the bigger picture.  The real parsing of this line is in the subsequent 'sscanf()'.  The '/proc' man page indicates that that the whole line is intended to be scanf-parseable.  This means that my 'strrchr()' to skip the early part can be subsumed into that 'sscanf()' which eliminates this part of my code altogether.  See 'eb874bd' below.